File: OpenTokFs.RequestTypes/OpenTokBroadcastStartRequest.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;

namespace OpenTokFs.RequestTypes {
    [JsonConverter(typeof(OpenTokBroadcastStartRequestPropertyConverter))]
    public class OpenTokBroadcastStartRequest {
        public string SessionId { get; }

        public OpenTokVideoLayout Layout { get; set; } = OpenTokVideoLayout.BestFit;
        public TimeSpan Duration { get; set; } = TimeSpan.FromHours(2);
        public bool Hls { get; set; } = false;
        public IEnumerable<OpenTokRtmpDestination> Rtmp { get; set; } = Enumerable.Empty<OpenTokRtmpDestination>();
        public string Resolution { get; set; } = "640x480";

        public OpenTokBroadcastStartRequest(string sessionId) {
            this.SessionId = sessionId ?? throw new ArgumentNullException(nameof(sessionId));
        }
    }

    public class OpenTokBroadcastStartRequestPropertyConverter : JsonConverter {
        public override bool CanConvert(Type objectType) {
            return objectType == typeof(OpenTokBroadcastStartRequest);
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer) {
            return false;
        }

        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) {
            if (value == null) {
                serializer.Serialize(writer, null);
                return;
            }

            var obj = value as OpenTokBroadcastStartRequest;

            var outputs = new Dictionary<string, object>();
            if (obj.Hls)
                outputs.Add("hls", new object());
            outputs.Add("rtmp", obj.Rtmp);

            var dict = new Dictionary<string, object> {
                ["sessionId"] = obj.SessionId,
                ["layout"] = obj.Layout,
                ["maxDuration"] = (int)obj.Duration.TotalSeconds,
                ["outputs"] = outputs,
                ["resolution"] = obj.Resolution
            };
            serializer.Serialize(writer, dict);
        }
    }
}

File: OpenTokFs/RequestTypes/OpenTokBroadcastStartRequest.fs
namespace OpenTokFs.RequestTypes

open System
open Newtonsoft.Json

[<JsonConverter(typeof<OpenTokBroadcastStartRequestPropertyConverter>)>]
type OpenTokBroadcastStartRequest(sessionId: string) =
    member __.sessionId = sessionId

    member val layout: OpenTokVideoLayout = OpenTokVideoLayout.BestFit with get, set
    member val duration: TimeSpan = TimeSpan.FromHours 2 with get, set
    member val Hls: bool = false with get, set
    member val Rtmp: OpenTokRtmpDestination seq = Seq.empty with get, set
    member val resolution: string = "640x480" with get, set

and OpenTokBroadcastStartRequestPropertyConverter() =
    inherit JsonConverter()

    override _.CanConvert t =
        t = typeof<OpenTokBroadcastStartRequest>

    override _.ReadJson (reader, objectType, existingValue, serializer) =
        base.ReadJson (reader, objectType, existingValue, serializer)

    override _.WriteJson (writer, value, serializer) =
        match value with
        | :? OpenTokBroadcastStartRequest as obj ->
            serializer.Serialize (writer, dict [
                ("sessionId", obj.sessionId :> obj)
                ("layout", obj.layout :> obj)
                ("maxDuration", obj.duration.TotalSeconds :> obj)
                ("outputs", [
                    if obj.Hls then ("hls", {||} :> obj)
                    ("rtmp", obj.Rtmp :> obj)
                ] :> obj)
                ("resolution", obj.resolution :> obj)
            ])
        | _ ->
            serializer.Serialize (writer, value)
